README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# MOR

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/Lifecycle-Stable-97ca00)](https://lifecycle.r-lib.org/articles/stages.html#stable)
![CRAN status](https://img.shields.io/badge/CRAN-not%20published-red)
[![test-coverage](https://github.com/shafayetShafee/MOR/actions/workflows/test-coverage.yaml/badge.svg)](https://github.com/shafayetShafee/MOR/actions/workflows/test-coverage.yaml)
[![R-CMD-check](https://github.com/shafayetShafee/MOR/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/shafayetShafee/MOR/actions/workflows/R-CMD-check.yaml)
[![Codecov test
coverage](https://codecov.io/gh/shafayetShafee/MOR/branch/main/graph/badge.svg)](https://app.codecov.io/gh/shafayetShafee/MOR?branch=main)
<!-- badges: end -->

## Overview

The goal of `MOR` is to provides post-estimation function to calculate
the Median Odds Ratio (MOR) from a multilevel binary logistic regression
model fit. Currently this package provides the median odds ratio in case
of two-level random intercept model and three-level random intercept
model only where the model is fitted using the R packages
[`lme4`](https://github.com/lme4/lme4/) or
[`GLMMadaptive`](https://drizopoulos.github.io/GLMMadaptive/index.html)
or [`glmmTMB`](https://glmmtmb.github.io/glmmTMB/).

## Installation

`{MOR}` is not on CRAN yet. You can install the development version of
MOR from [GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
remotes::install_github("shafayetShafee/MOR")
```

## Example

``` r
library(MOR)

data("mlm_data1", package = "MOR")
data("mlm_data2", package = "MOR")

# fitting two level random intercept model using lme4 package
model <- lme4::glmer(Yij ~ X1c + X2b + (1 | cluster),
  family = "binomial", data = mlm_data1
)

mor(model)
#> # A tibble: 1 × 4
#>   term        estimate ci_lower ci_upper
#>   <chr>          <dbl>    <dbl>    <dbl>
#> 1 mor_cluster     4.42     3.51     5.58
## to get 90% CI
mor(model, conf.level = 0.90)
#> # A tibble: 1 × 4
#>   term        estimate ci_lower ci_upper
#>   <chr>          <dbl>    <dbl>    <dbl>
#> 1 mor_cluster     4.42     3.64     5.38

# fitting two level random intercept model using GLMMadaptive package
model1 <- GLMMadaptive::mixed_model(
  fixed = Yij ~ X1c + X2b,
  random = ~ 1 | cluster,
  family = binomial("logit"), data = mlm_data1
)

mor(model1)
#> # A tibble: 1 × 4
#>   term        estimate ci_lower ci_upper
#>   <chr>          <dbl>    <dbl>    <dbl>
#> 1 mor_cluster     4.43     3.51     5.60


# fitting two level random intercept model using glmmTMB package
model2 <- glmmTMB::glmmTMB(Yij ~ X1c + X2b + (1 | cluster),
  family = binomial("logit"), data = mlm_data1
)

mor(model2)
#> # A tibble: 1 × 4
#>   term        estimate ci_lower ci_upper
#>   <chr>          <dbl>    <dbl>    <dbl>
#> 1 mor_cluster     4.42     3.50     5.59


# fitting three level random intercept model using glmmTMB package
model3 <- glmmTMB::glmmTMB(Yijk ~ X1c + X2b + (1 | ea) + (1 | ea:hh),
  family = "binomial", data = mlm_data2
)

mor(model3)
#> # A tibble: 2 × 4
#>   term      estimate ci_lower ci_upper
#>   <chr>        <dbl>    <dbl>    <dbl>
#> 1 mor_ea:hh     4.21     3.78     4.68
#> 2 mor_ea        6.29     5.38     7.36


# fitting three level random intercept model using lme4 package
model4 <- lme4::glmer(Yijk ~ X1c + X2b + (1 | ea) + (1 | ea:hh),
  family = "binomial", data = mlm_data2
)

mor(model4)
#> # A tibble: 2 × 4
#>   term      estimate ci_lower ci_upper
#>   <chr>        <dbl>    <dbl>    <dbl>
#> 1 mor_ea:hh     4.21     3.78     4.68
#> 2 mor_ea        6.29     5.38     7.36
## to get 90% CI
mor(model4, conf.level = 0.90)
#> # A tibble: 2 × 4
#>   term      estimate ci_lower ci_upper
#>   <chr>        <dbl>    <dbl>    <dbl>
#> 1 mor_ea:hh     4.21     3.85     4.60
#> 2 mor_ea        6.29     5.52     7.18
```

## Code of Conduct

Please note that the MOR project is released with a [Contributor Code of
Conduct](https://contributor-covenant.org/version/2/1/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# MOR 0.2.3

* Removed the argument to get `SE` of `MOR`, since the there was mistakes in implementation to get `SE` of `MOR`, but afraid not, `CI` was correctly calculated using the Delta method.

# MOR 0.2.2

* Done some internal changes

# MOR 0.2.1

* Fixed a bug in making `NULL` when `conf.int = FALSE`

# MOR 0.2.0

* Added method for `glmerMod` object from `lme4::glmer`
* Added a bit info on MOR and references in function doc

# MOR 0.1.0

* Created the package.
* Created the `mor` generic function.
* Added `mor` method for `MixMod` (model object from [`GLMMadaptive`](https://drizopoulos.github.io/GLMMadaptive/index.html) package)
* Added `mor` method for `glmmTMB` (model object from [`glmmTMB`](https://glmmtmb.github.io/glmmTMB/) package)



tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(MOR)

test_check("MOR")


tests/testthat/test-mor.R
test_that("Expecting Errors for fitting models with family other than binomial", {
  # taken from
  # https://drizopoulos.github.io/GLMMadaptive/articles/GLMMadaptive.html
  set.seed(1234)
  n <- 100
  K <- 8
  t_max <- 15

  DF <- data.frame(
    id = rep(seq_len(n), each = K),
    time = c(replicate(n, c(0, sort(runif(K - 1, 0, t_max))))),
    sex = rep(gl(2, n / 2, labels = c("male", "female")), each = K)
  )

  X <- model.matrix(~ sex * time, data = DF)

  betas <- c(2.13, -0.25, 0.24, -0.05)
  D11 <- 0.48

  b <- rnorm(n, sd = sqrt(D11))
  eta_y <- drop(X %*% betas + b[DF$id])
  DF$y <- rpois(n * K, exp(eta_y))

  gm1 <- GLMMadaptive::mixed_model(
    fixed = y ~ sex * time, random = ~ 1 | id, data = DF,
    family = poisson()
  )
  expect_error(
    mor(gm1),
    "MOR can only be calculated for Multilevel Logistic Regression Model i.e. for `MixMod` with `binomial` family with `logit` link"
  )
})


test_that("Expecting Errors for fitting two level random slope model using GLMMadaptive", {
  data("mlm_data1")

  # fitting two level random slope model using GLMMadaptive package
  model1 <- GLMMadaptive::mixed_model(
    fixed = Yij ~ X1c + X2b,
    random = ~ X1c | cluster,
    family = binomial("logit"), data = mlm_data1
  )

  expect_error(
    mor(model1),
    "MOR can only be calculated for Two level random intercept model."
  )
})


test_that("Expecting Errors for fitting two level random slope model using glmmTMB", {
  data("mlm_data1")

  # fitting two level random slope model using glmmTMB package
  model2 <- glmmTMB::glmmTMB(Yij ~ X1c + X2b + (X1c | cluster),
    family = binomial("logit"), data = mlm_data1
  )

  expect_error(
    mor(model2),
    "MOR can only be calculated for Two level random intercept model."
  )
})


test_that("Expecting Errors for fitting three level random slope model using glmmTMB", {
  data("mlm_data2")

  # fitting two level random slope model using glmmTMB package
  model3 <- glmmTMB::glmmTMB(Yijk ~ X1c + X2b + (X1c | ea:hh) + (1 | ea),
    family = "binomial", data = mlm_data2
  )

  expect_error(
    mor(model3),
    "MOR can only be calculated for Three level random intercept model."
  )
})


test_that("Expecting Equal results for fitting three level random slope model using glmmTMB using different style", {
  data("mlm_data2")

  # fitting two level random slope model using glmmTMB package
  model4 <- glmmTMB::glmmTMB(Yijk ~ X1c + X2b + (1 | ea:hh) + (1 | ea),
    family = "binomial", data = mlm_data2
  )

  model5 <- glmmTMB::glmmTMB(Yijk ~ X1c + X2b + (1 | ea) + (1 | ea:hh),
    family = "binomial", data = mlm_data2
  )

  expect_equal(mor(model4), mor(model5))
})


test_that("Expecting Errors for fitting two level random slope model using lme4", {
  data("mlm_data1")

  # fitting two level random slope model using glmmTMB package
  model6 <- lme4::glmer(Yij ~ X1c + X2b + (X1c | cluster),
    family = binomial("logit"), data = mlm_data1
  )

  expect_error(
    mor(model6),
    "MOR can only be calculated for Two level random intercept model."
  )
})


test_that("Expecting Errors for fitting three level random slope model using lme4", {
  data("mlm_data2")

  # fitting two level random slope model using glmmTMB package
  model7 <- lme4::glmer(Yijk ~ X1c + X2b + (X1c | ea:hh) + (1 | ea),
    family = "binomial", data = mlm_data2
  )

  expect_error(
    mor(model7),
    "MOR can only be calculated for Three level random intercept model."
  )
})


test_that("Expecting Equal results for fitting three level random slope model using lme4 using different style", {
  data("mlm_data2")

  # fitting two level random slope model using lme4 package
  model8 <- lme4::glmer(Yijk ~ X1c + X2b + (1 | ea:hh) + (1 | ea),
    family = "binomial", data = mlm_data2
  )

  model9 <- lme4::glmer(Yijk ~ X1c + X2b + (1 | ea) + (1 | ea:hh),
    family = "binomial", data = mlm_data2
  )

  expect_equal(mor(model8), mor(model9))
})


test_that("Expecting Errors for mor.default", {
  data("mlm_data1")

  # fitting two level random slope model using glmmTMB package
  model5 <- stats::glm(Yij ~ X1c + X2b, family = "binomial", data = mlm_data1)

  expect_error(
    mor(model5),
    "`mor` does not work for models of class *."
  )

  expect_error(
    mor(1:3),
    "`mor` does not work for models of class *."
  )
})


test_that("Check for equivalent answer of mor for different package fit (two level int)", {
  data("mlm_data1")

  model10 <- lme4::glmer(Yij ~ X1c + X2b + (1 | cluster),
    family = "binomial",
    data = mlm_data1
  )

  model11 <- GLMMadaptive::mixed_model(
    fixed = Yij ~ X1c + X2b,
    random = ~ 1 | cluster,
    family = "binomial", data = mlm_data1
  )


  model12 <- glmmTMB::glmmTMB(Yij ~ X1c + X2b + (1 | cluster),
    family = "binomial",
    data = mlm_data1
  )

  expect_equal(mor(model10), mor(model11), tolerance = 0.1)
  expect_equal(mor(model10), mor(model11), tolerance = 0.1)
  expect_equal(mor(model11), mor(model12), tolerance = 0.1)
})


test_that("Check for equivalent answer of mor for different package fit (three level int)", {
  data("mlm_data2")

  model13 <- lme4::glmer(Yijk ~ X1c + X2b + (1 | ea:hh) + (1 | ea),
    family = "binomial",
    data = mlm_data2
  )

  model14 <- glmmTMB::glmmTMB(Yijk ~ X1c + X2b + (1 | ea:hh) + (1 | ea),
    family = "binomial",
    data = mlm_data2
  )

  expect_equal(mor(model13), mor(model14), tolerance = 0.1)
})


test_that("Testing for valid arguments", {
  data("mlm_data1")

  model15 <- lme4::glmer(Yij ~ X1c + X2b + (1 | cluster),
    family = "binomial", data = mlm_data1
  )

  expect_error(
    mor(model15, conf.int = NULL),
    "`conf.int` must be a boolean i.e. `TRUE` or `FALSE`"
  )

  expect_error(
    mor(model15, conf.int = NA),
    "`conf.int` must be a boolean i.e. `TRUE` or `FALSE`"
  )

  expect_error(
    mor(model15, conf.int = NaN),
    "`conf.int` must be a boolean i.e. `TRUE` or `FALSE`"
  )

  expect_error(
    mor(model15, conf.int = 1),
    "`conf.int` must be a boolean i.e. `TRUE` or `FALSE`"
  )

  expect_error(
    mor(model15, conf.int = "1"),
    "`conf.int` must be a boolean i.e. `TRUE` or `FALSE`"
  )

  expect_error(
    mor(model15, conf.level = 1),
    "`conf.level` must be a numeric less than 1 and greated than 0"
  )

  expect_error(
    mor(model15, conf.level = 0),
    "`conf.level` must be a numeric less than 1 and greated than 0"
  )

  expect_error(
    mor(model15, conf.level = 95),
    "`conf.level` must be a numeric less than 1 and greated than 0"
  )

  expect_error(
    mor(model15, conf.level = NULL),
    "`conf.level` must be a numeric less than 1 and greated than 0"
  )

  expect_error(
    mor(model15, conf.level = NA),
    "`conf.level` must be a numeric less than 1 and greated than 0"
  )

  expect_error(
    mor(model15, conf.level = NaN),
    "`conf.level` must be a numeric less than 1 and greated than 0"
  )
})


R/mor.R
#' Calculate Median Odds Ratio from a fitted multilevel binary logistic model object.
#'
#' @param object An `glmerMod` object created by [`lme4::glmer()`] or an `MixMod` object
#'   created by [`GLMMadaptive::mixed_model()`] or
#'   an `glmmTMB` object created by [`glmmTMB::glmmTMB()`]. See Details.
#' @param conf.int Logical indicating whether or not to include a confidence
#'   interval using the Delta method in the tidied output. Defaults to `TRUE`.
#' @param conf.level The confidence level to use for the confidence interval
#'  if `conf.int = TRUE`. Must be strictly greater than 0 and less than 1.
#'  Defaults to 0.95, which corresponds to a 95 percent confidence interval.
#' @param ... Currently not used.
#'
#' @details
#' Median Odds Ratio (MOR) is suggested as a measure of heterogeneity that quantifies
#' cluster heterogeneity and facilitates a direct comparison between covariate effects
#' and the magnitude of heterogeneity in terms of well-known odds ratios
#' \insertCite{larsen2000interpreting,larsen2005appropriate}{MOR}
#'
#' @return a [tibble][tibble::tibble-package] with columns,
#'  * `term`:      Name of the estimate
#'  * `estimate`:  Estimate of Median Odds Ratio (MOR)
#'  * `ci_lower`:  Lower bound of the confidence interval of MOR
#'  * `ci_upper`:  Upper bound of the confidence interval of MOR
#'
#' @references
#' \insertAllCited{}
#'
#' @examples
#' data("mlm_data1")
#' data("mlm_data2")
#'
#' # fitting two level random intercept model using lme4 package
#' model <- lme4::glmer(Yij ~ X1c + X2b + (1 | cluster),
#'   family = "binomial", data = mlm_data1
#' )
#'
#' mor(model)
#' ## to get 90% CI
#' mor(model, conf.level = 0.90)
#'
#' # fitting two level random intercept model using GLMMadaptive package
#' model1 <- GLMMadaptive::mixed_model(
#'   fixed = Yij ~ X1c + X2b,
#'   random = ~ 1 | cluster,
#'   family = binomial("logit"),
#'   data = mlm_data1
#' )
#'
#' mor(model1)
#'
#' # fitting two level random intercept model using glmmTMB package
#' model2 <- glmmTMB::glmmTMB(Yij ~ X1c + X2b + (1 | cluster),
#'   family = binomial("logit"), data = mlm_data1
#' )
#'
#' mor(model2)
#'
#' # fitting three level random intercept model using glmmTMB package
#' model3 <- glmmTMB::glmmTMB(Yijk ~ X1c + X2b + (1 | ea) + (1 | ea:hh),
#'   family = "binomial", data = mlm_data2
#' )
#'
#' mor(model3)
#'
#' # fitting three level random intercept model using lme4 package
#' model4 <- lme4::glmer(Yijk ~ X1c + X2b + (1 | ea) + (1 | ea:hh),
#'   family = "binomial", data = mlm_data2
#' )
#'
#'
#' mor(model4)
#'
#' @importFrom Rdpack reprompt
#' @export
mor <- function(object, conf.int = TRUE, conf.level = 0.95, ...) {
  check_args(conf.int, conf.level)
  UseMethod("mor")
}


#' @export
mor.default <- function(object, ...) {
  err_msg <- paste0(
    sprintf("`mor` does not work for models of class '%s'.", class(object)[1]),
    "It currently supports `MixMod` object (fitted model object) from `GLMMadaptive` package and `glmmTMB` object from `glmmTMB` package"
  )
  stop(err_msg, call. = FALSE)
}


#' @importFrom stats qnorm
#' @export
mor.MixMod <- function(object, conf.int = TRUE, conf.level = 0.95,
                       ranef.info = FALSE, ...) {
  obj_family <- object$family
  if (obj_family$family != "binomial" && obj_family$link != "logit") {
    stop("MOR can only be calculated for Multilevel Logistic Regression Model i.e. for `MixMod` with `binomial` family with `logit` link",
      call. = FALSE
    )
  }

  if (!all(dim(object$D) == 1)) {
    stop("MOR can only be calculated for Two level random intercept model.", call. = FALSE)
  }

  grp_var_name <- object$id_name

  sigma_u_sq_hat <- object$D[[1]]
  var_sigma_u_sq_hat <- vcov_orig_scale(object)

  mor_hat <- exp(sqrt(2 * sigma_u_sq_hat) * qnorm(0.75))
  log_mor_hat <- log(mor_hat)

  log_mor_int_expr <- function(x) sqrt(2 * x) * qnorm(0.75)
  J <- numDeriv::jacobian(log_mor_int_expr, x = sigma_u_sq_hat)
  log_se_mor_hat <- as.numeric(sqrt(t(J) %*% var_sigma_u_sq_hat %*% J))
  se_mor_hat <- exp(log_se_mor_hat)

  crit_val <- qnorm((1 - conf.level) / 2, lower.tail = FALSE)
  ci <- log_mor_hat + c(-1, 1) * crit_val * log_se_mor_hat
  ci_exp <- exp(ci)
  ci_lower <- ci_exp[1]
  ci_upper <- ci_exp[2]

  if (!conf.int) {
    ci_lower <- NULL
    ci_upper <- NULL
  }

  # ranef info
  ranef <- unname(sigma_u_sq_hat)
  se_ranef <- unname(sqrt(as.numeric(var_sigma_u_sq_hat)))

  if (!ranef.info) {
    ranef <- NULL
    se_ranef <- NULL
  }

  return(tibble::tibble(
    term = paste0("mor_", grp_var_name),
    estimate = mor_hat,
    ci_lower = ci_lower,
    ci_upper = ci_upper,
    ranef_var_comp = ranef,
    se_var_comp = se_ranef
  ))
}


#' @importFrom stats qnorm vcov
#' @export
mor.glmmTMB <- function(object, conf.int = TRUE, conf.level = 0.95,
                        ranef.info = FALSE, ...) {
  obj_family <- object$modelInfo$family
  if (obj_family$family != "binomial" && obj_family$link != "logit") {
    stop("MOR can only be calculated for Multilevel Binary Logistic Regression Model i.e. for `MixMod` with `binomial` family with `logit` link",
      call. = FALSE
    )
  }

  par_vcov <- diag(vcov(object, full = TRUE))
  par_names <- names(par_vcov)
  grp_var_names <- object$modelInfo$grpVar
  grp_var_no <- length(grp_var_names)
  ran_par_no <- sum(grepl("theta", par_names))

  pattern <- paste0("theta.*[", paste0(grp_var_names, collapse = "|"), "]")
  ran_terms_idx <- which(grepl(pattern, par_names))

  se_logsd <- sqrt(par_vcov)[ran_terms_idx]
  logsd <- object$sdr$par.fixed[ran_terms_idx]
  sigma_sq_hat <- unname(exp(2 * logsd))
  # by delta method
  se_sigma_sq_hat <- se_logsd * 2 * exp(2 * logsd)
  var_sigma_sq_hat <- se_sigma_sq_hat^2

  if (grp_var_no == 1) {
    if (ran_par_no > 1) {
      stop("MOR can only be calculated for Two level random intercept model.", call. = FALSE)
    }

    mor_hat <- exp(sqrt(2 * sigma_sq_hat) * qnorm(0.75))
    log_mor_hat <- log(mor_hat)

    log_mor_expr <- function(x) {
      # x is ran-effect parameterized as sd
      sqrt(2 * x) * qnorm(0.75)
    }

    J <- numDeriv::jacobian(log_mor_expr, x = sigma_sq_hat)
    log_se_mor_hat <- as.numeric(sqrt(t(J) %*% var_sigma_sq_hat %*% J))
    se_mor_hat <- exp(log_se_mor_hat)

    crit_val <- stats::qnorm((1 - conf.level) / 2, lower.tail = FALSE)
    ci <- log_mor_hat + c(-1, 1) * crit_val * log_se_mor_hat
    ci_exp <- exp(ci)
    ci_lower <- ci_exp[1]
    ci_upper <- ci_exp[2]

    if (!conf.int) {
      ci_lower <- NULL
      ci_upper <- NULL
    }

    # ranef info
    ranef <- unname(sigma_sq_hat)
    se_ranef <- unname(sqrt(as.numeric(var_sigma_sq_hat)))

    if (!ranef.info) {
      ranef <- NULL
      se_ranef <- NULL
    }

    return(tibble::tibble(
      term = paste0("mor_", grp_var_names),
      estimate = mor_hat,
      ci_lower = ci_lower,
      ci_upper = ci_upper,
      ranef_var_comp = ranef,
      se_var_comp = se_ranef
    ))
  } else if (grp_var_no == 2) {
    if (!ran_par_no == 2) {
      stop("MOR can only be calculated for Three level random intercept model.", call. = FALSE)
    }

    fix_terms_names <- names(attr(object$modelInfo$terms$cond$fixed, "dataClasses"))
    model_terms_names <- names(object$frame)
    diff <- setdiff(model_terms_names, fix_terms_names)
    possible_nested_term <- c(
      paste0(diff, collapse = ":"),
      paste0(rev(diff), collapse = ":")
    )
    nested_term_names <- intersect(grp_var_names, possible_nested_term)
    third_lvl_var_names <- setdiff(grp_var_names, nested_term_names)

    sigma_ujk_idx <- grepl(nested_term_names, names(se_sigma_sq_hat), fixed = TRUE)
    sigma_ujk <- sigma_sq_hat[sigma_ujk_idx]
    mor1_hat <- exp(sqrt(2 * sigma_ujk) * qnorm(0.75))
    log_mor1_hat <- log(mor1_hat)

    log_mor1_expr <- function(x) {
      # x is ran-effect parameterized as sd
      sqrt(2 * x) * qnorm(0.75)
    }

    J1 <- numDeriv::jacobian(log_mor1_expr, x = sigma_ujk)
    log_se_mor1_hat <- as.numeric(sqrt(t(J1) %*% var_sigma_sq_hat[sigma_ujk_idx] %*% J1))
    se_mor1_hat <- exp(log_se_mor1_hat)

    # mor2 calc. ----------------------
    mor2_hat <- exp(sqrt(2 * sum(sigma_sq_hat)) * qnorm(0.75))
    log_mor2_hat <- log(mor2_hat)

    log_mor2_expr <- function(x) {
      # x is ran-effect parameterized as variance
      sqrt(2 * sum(x)) * qnorm(0.75)
    }

    J2 <- numDeriv::jacobian(log_mor2_expr, x = sigma_sq_hat)
    log_se_mor2_hat <- as.numeric(sqrt(J2 %*% diag(var_sigma_sq_hat) %*% t(J2)))
    se_mor2_hat <- exp(log_se_mor2_hat)

    crit_val <- stats::qnorm((1 - conf.level) / 2, lower.tail = FALSE)

    ci_1 <- log_mor1_hat + c(-1, 1) * crit_val * log_se_mor1_hat
    ci_1_exp <- exp(ci_1)
    ci_lower_1 <- ci_1_exp[1]
    ci_upper_1 <- ci_1_exp[2]

    ci_2 <- log_mor2_hat + c(-1, 1) * crit_val * log_se_mor2_hat
    ci_2_exp <- exp(ci_2)
    ci_lower_2 <- ci_2_exp[1]
    ci_upper_2 <- ci_2_exp[2]

    if (!conf.int) {
      ci_lower_1 <- NULL
      ci_upper_1 <- NULL
      ci_lower_2 <- NULL
      ci_upper_2 <- NULL
    }

    # ranef info
    ranef <- unname(c(sigma_sq_hat[sigma_ujk_idx], sigma_sq_hat[!sigma_ujk_idx]))
    se_ranef <- unname(sqrt(
      c(var_sigma_sq_hat[sigma_ujk_idx], var_sigma_sq_hat[!sigma_ujk_idx])
    ))

    if (!ranef.info) {
      ranef <- NULL
      se_ranef <- NULL
    }

    return(tibble::tibble(
      term = c(paste0("mor_", nested_term_names), paste0("mor_", third_lvl_var_names)),
      estimate = c(mor1_hat, mor2_hat),
      ci_lower = c(ci_lower_1, ci_lower_2),
      ci_upper = c(ci_upper_1, ci_upper_2),
      ranef_var_comp = ranef,
      se_var_comp = se_ranef
    ))
  }
}


#' @importFrom stats qnorm vcov
#' @export
mor.glmerMod <- function(object, conf.int = TRUE, conf.level = 0.95,
                         ranef.info = FALSE, ...) {
  s <- summary(object)
  if (s$family != "binomial" && s$link != "logit") {
    stop("MOR can only be calculated for Multilevel Logistic Regression Model i.e. for `glmerMod` with `binomial` family with `logit` link",
      call. = FALSE
    )
  }

  ran_eff_df <- as.data.frame(s$varcor)
  grp_var_no <- length(s$ngrps)
  grp_var_name <- unique(ran_eff_df$grp)

  if (grp_var_no == 1) {
    if (nrow(ran_eff_df) > 1) {
      stop("MOR can only be calculated for Two level random intercept model.", call. = FALSE)
    }

    grp_var_name_pat <- paste0("cov_", grp_var_name, ".\\(Intercept\\)")
    sigma_u_sq <- ran_eff_df$vcov
    vcov_mat <- merDeriv::vcov.glmerMod(object, full = TRUE, ranpar = "sd")
    idx <- which(grepl(grp_var_name_pat, colnames(vcov_mat)))
    se_sigma_u <- sqrt(diag(vcov_mat)[idx])

    mor_hat <- exp(sqrt(2 * sigma_u_sq) * qnorm(0.75))
    log_mor_hat <- log(mor_hat)
    log_se_mor_hat <- sqrt(2) * qnorm(0.75) * se_sigma_u
    se_mor_hat <- exp(log_se_mor_hat)

    crit_val <- qnorm((1 - conf.level) / 2, lower.tail = FALSE)
    ci <- log_mor_hat + c(-1, 1) * crit_val * log_se_mor_hat
    ci_exp <- exp(ci)
    ci_lower <- ci_exp[1]
    ci_upper <- ci_exp[2]

    if (!conf.int) {
      ci_lower <- NULL
      ci_upper <- NULL
    }

    # ranef info
    ranef <- unname(sqrt(sigma_u_sq))
    se_ranef <- unname(as.numeric(se_sigma_u))

    if (!ranef.info) {
      ranef <- NULL
      se_ranef <- NULL
    }

    return(tibble::tibble(
      term = paste0("mor_", grp_var_name),
      estimate = mor_hat,
      ci_lower = ci_lower,
      ci_upper = ci_upper,
      ranef_sd_comp = ranef,
      se_sd_comp = se_ranef
    ))
  } else if (grp_var_no == 2) {
    if (nrow(ran_eff_df) > 2) {
      stop("MOR can only be calculated for Three level random intercept model.", call. = FALSE)
    }

    sigma_hat <- ran_eff_df$sdcor
    var_sigma_hat <- diag(diag(2 * solve(object@optinfo$derivs$Hessian))[1:2])

    mor1_hat <- exp(sqrt(2 * sigma_hat[1]^2) * qnorm(0.75))
    log_mor1_hat <- log(mor1_hat)
    log_mor1_expr <- function(x) sqrt(2 * x^2) * qnorm(0.75)
    J <- numDeriv::jacobian(log_mor1_expr, x = sigma_hat[1])
    log_se_mor1_hat <- as.numeric(sqrt(t(J) %*% var_sigma_hat[1, 1] %*% J))
    se_mor1_hat <- exp(log_se_mor1_hat)

    mor2_hat <- exp(sqrt(2 * sum(sigma_hat^2)) * qnorm(0.75))
    log_mor2_hat <- log(mor2_hat)
    log_mor2_expr <- function(x) sqrt(2 * sum(x^2)) * qnorm(0.75)
    J2 <- numDeriv::jacobian(log_mor2_expr, x = sigma_hat)
    log_se_mor2_hat <- as.numeric(sqrt(J2 %*% var_sigma_hat %*% t(J2)))
    se_mor2_hat <- exp(log_se_mor2_hat)

    crit_val <- stats::qnorm((1 - conf.level) / 2, lower.tail = FALSE)

    ci_1 <- log_mor1_hat + c(-1, 1) * crit_val * log_se_mor1_hat
    ci_1_exp <- exp(ci_1)
    ci_lower_1 <- ci_1_exp[1]
    ci_upper_1 <- ci_1_exp[2]

    ci_2 <- log_mor2_hat + c(-1, 1) * crit_val * log_se_mor2_hat
    ci_2_exp <- exp(ci_2)
    ci_lower_2 <- ci_2_exp[1]
    ci_upper_2 <- ci_2_exp[2]

    if (!conf.int) {
      ci_lower_1 <- NULL
      ci_upper_1 <- NULL
      ci_lower_2 <- NULL
      ci_upper_2 <- NULL
    }

    # ranef info
    ranef <- unname(sigma_hat)
    se_ranef <- unname(sqrt(as.numeric(
      diag(var_sigma_hat)
    )))

    if (!ranef.info) {
      ranef <- NULL
      se_ranef <- NULL
    }

    return(tibble::tibble(
      term = c(paste0("mor_", grp_var_name[1]), paste0("mor_", grp_var_name[2])),
      estimate = c(mor1_hat, mor2_hat),
      ci_lower = c(ci_lower_1, ci_lower_2),
      ci_upper = c(ci_upper_1, ci_upper_2),
      ranef_sd_comp = ranef,
      se_sd_comp = se_ranef
    ))
  }
}





# helper funs -------------------------------------------------------------

# taken from GLMMadaptive pkg
chol_transf <- function(x) {
  if (any(is.na(x) | !is.finite(x))) {
    stop("NA or infinite values in 'x'.\n")
  }
  if (is.matrix(x)) {
    k <- nrow(x)
    U <- chol(x)
    U[cbind(1:k, 1:k)] <- log(U[cbind(1:k, 1:k)])
    U[upper.tri(U, TRUE)]
  } else {
    nx <- length(x)
    k <- round((-1 + sqrt(1 + 8 * nx)) / 2)
    mat <- matrix(0, k, k)
    mat[upper.tri(mat, TRUE)] <- x
    mat[cbind(1:k, 1:k)] <- exp(mat[cbind(1:k, 1:k)])
    res <- crossprod(mat)
    attr(res, "L") <- t(mat)[lower.tri(mat, TRUE)]
    res
  }
}


# copied from https://stackoverflow.com/a/76925070/10858321 by @statmerkur
D_chol_to_D <- function(x) {
  # transform the log-cholesky parameterized value back to original scale
  D <- chol_transf(x)
  D[upper.tri(D, diag = TRUE)]
}


#' @importFrom stats vcov
vcov_orig_scale <- function(model) {
  D <- model$D
  D_chol_entries <- chol_transf(D)
  V_chol <- vcov(model, parm = "var-cov")
  J <- numDeriv::jacobian(D_chol_to_D, D_chol_entries)
  V <- J %*% V_chol %*% t(J)
  colnames(V) <- colnames(V_chol)
  rownames(V) <- rownames(V_chol)
  return(V)
}


is_bool <- function(x) {
  is.logical(x) && length(x) == 1L && !is.na(x)
}


check_logical <- function(x) {
  if (invalid(x) || !is_bool(x)) {
    stop(
      paste0("`", substitute(x), "` must be a boolean i.e. `TRUE` or `FALSE`"),
      call. = FALSE
    )
  }
}


check_args <- function(conf.int, conf.level) {
  check_logical(conf.int)
  if (invalid(conf.level) || !(conf.level > 0 && conf.level < 1)) {
    stop("`conf.level` must be a numeric less than 1 and greated than 0", call. = FALSE)
  }
}


# copied from https://github.com/r-gregmisc/gtools/blob/master/R/invalid.R
#' @importFrom methods is
invalid <- function(x) {
  if (missing(x) || is.null(x) || length(x) == 0 || is(x, "try-error")) {
    return(TRUE)
  }
  if (is.list(x)) {
    return(all(sapply(x, invalid)))
  } else if (is.vector(x)) {
    return(all(is.na(x)))
  } else {
    return(FALSE)
  }
}
